File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj_list = defaultdict(set)
        indegree = [0 for _ in range(numCourses)]

        for course, pre_req in prerequisites:
            adj_list[pre_req].add(course)
            indegree[course] += 1
        
        result = []

        q = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)
                result.append(i)

        while q:
            curr_course = q.popleft()
            for course in adj_list[curr_course]:
                indegree[course] -= 1
                if indegree[course] == 0:
                    q.append(course)
                    result.append(course)


        return result if len(result) == numCourses else []


        

File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> adj_list;
        vector<int> in_degree(numCourses, 0);

        for (const auto& prereq : prerequisites) {
            int course = prereq[0];
            int prerequisite = prereq[1];
            adj_list[prerequisite].push_back(course);
            in_degree[course]++;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; ++i) {
            if (in_degree[i] == 0) {
                q.push(i);
            }
        }

        vector<int> order;
        while (!q.empty()) {
            int current = q.front();
            q.pop();
            order.push_back(current);

            for (int neighbor : adj_list[current]) {
                in_degree[neighbor]--;
                if (in_degree[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }

        if (order.size() == numCourses) {
            return order;
        } else {
            return {};
        }
    }
};
